File: mobile/src/app/pages/client/settings/client-settings.page.ts
import { DatePipe } from '@angular/common';
import { Component, OnInit, inject } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AlertController, ToastController } from '@ionic/angular';
import {
  IonBackButton,
  IonButton,
  IonButtons,
  IonContent,
  IonHeader,
  IonTitle,
  IonToolbar
} from '@ionic/angular/standalone';

import {
  ClientDetailChangeRequest,
  DynamicField
} from '../../../core/api/forms-groups-api.service';
import { ClientApiService, ClientMeResponse } from '../../../core/api/client-api.service';
import { PasswordInputComponent } from '../../../shared/password-input.component';

/** Settings & Account — client information (edit via trainer-approved request), photo, password, deletion. */
@Component({
  selector: 'app-client-settings',
  standalone: true,
  imports: [DatePipe, FormsModule, PasswordInputComponent, IonHeader, IonToolbar, IonTitle, IonButtons, IonBackButton, IonButton, IonContent],
  template: `
    <ion-header>
      <ion-toolbar>
        <ion-buttons slot="start"><ion-back-button defaultHref="/client/tabs/more" /></ion-buttons>
        <ion-title>Settings &amp; Account</ion-title>
      </ion-toolbar>
    </ion-header>
    <ion-content>
      <div class="page-pad">
        @if (message) {
          <p [class]="messageIsError ? 'error-text' : 'success-text'">{{ message }}</p>
        }

        @if (me; as data) {
          <div class="profile-card">
            @if (data.client.photo) {
              <img class="avatar" [src]="data.client.photo" alt="" />
            } @else {
              <div class="avatar avatar-fallback">{{ initials }}</div>
            }
            <div style="flex:1;min-width:0">
              <strong>{{ data.client.first_name }} {{ data.client.last_name }}</strong>
              <small>{{ data.client.username }} · {{ data.client.group_name }}</small>
            </div>
          </div>

          <div class="card">
            <h3>Profile photo</h3>
            <input type="file" accept="image/*" (change)="onPhoto($event)" />
          </div>

          <div class="card">
            <div style="display:flex;justify-content:space-between;align-items:center">
              <h3 style="margin:0">Client information</h3>
              @if (!pendingRequest && !isEditing) {
                <ion-button size="small" fill="outline" (click)="startEdit()">Request edit</ion-button>
              }
            </div>

            @if (pendingRequest) {
              <p class="hint-note" style="margin-top:.5rem">
                Your edit request from {{ pendingRequest.created_at | date: 'dd MMM' }} is waiting for trainer review.
              </p>
            }

            @if (isEditing) {
              <div class="form-grid" style="margin-top:.7rem">
                @for (field of editableFields; track field.key) {
                  <label>
                    <span>{{ field.label }}</span>
                    <input [(ngModel)]="draftAnswers[field.key]" />
                  </label>
                }
                <label><span>Note to trainer (optional)</span><input [(ngModel)]="editNote" /></label>
              </div>
              <div class="form-two" style="margin-top:.7rem">
                <ion-button size="small" (click)="submitEdit()" [disabled]="isSubmittingEdit">
                  {{ isSubmittingEdit ? 'Sending…' : 'Send for approval' }}
                </ion-button>
                <ion-button size="small" fill="clear" (click)="isEditing = false">Cancel</ion-button>
              </div>
              <p class="hint-note">Changes take effect after your trainer approves them.</p>
            } @else {
              <div class="kv-list" style="margin-top:.6rem">
                <div class="kv"><span>Name</span><strong>{{ data.client.first_name }} {{ data.client.last_name }}</strong></div>
                <div class="kv"><span>Email</span><strong>{{ data.client.email || '—' }}</strong></div>
                <div class="kv"><span>Username</span><strong>{{ data.client.username }}</strong></div>
                <div class="kv"><span>Group</span><strong>{{ data.client.group_name }}</strong></div>
                <div class="kv"><span>Reference</span><strong>{{ data.client.reference_id }}</strong></div>
                <div class="kv"><span>Joined</span><strong>{{ data.client.created_at | date: 'dd MMM yyyy' }}</strong></div>
                <div class="kv"><span>Status</span><strong>{{ data.client.is_active ? 'Active' : 'Inactive' }}</strong></div>
                @for (field of answeredFields; track field.key) {
                  <div class="kv"><span>{{ field.label }}</span><strong>{{ field.value }}</strong></div>
                }
              </div>
            }
          </div>

          <div class="card">
            <h3>Change password</h3>
            <div class="form-grid">
              <label><span>Current password</span><app-password-input [(ngModel)]="currentPassword" /></label>
              <label><span>New password</span><app-password-input [(ngModel)]="newPassword" autocomplete="new-password" /></label>
              <label><span>Confirm new password</span><app-password-input [(ngModel)]="confirmPassword" autocomplete="new-password" /></label>
            </div>
            <ion-button size="small" style="margin-top:.6rem" (click)="changePassword()"
              [disabled]="isChangingPassword || !currentPassword || newPassword.length < 8 || newPassword !== confirmPassword">
              {{ isChangingPassword ? 'Updating…' : 'Update password' }}
            </ion-button>
          </div>

          <div class="card">
            <h3>Account deletion</h3>
            @if (deletionRequest && deletionRequest.status === 'pending') {
              <p class="sub">Your deletion request is waiting for trainer review.</p>
              <ion-button size="small" fill="outline" (click)="withdrawDeletion()">Withdraw request</ion-button>
            } @else {
              <p class="sub">Ask your trainer to delete your account and data. This requires their approval.</p>
              <ion-button size="small" color="danger" fill="outline" (click)="requestDeletion()">Request deletion</ion-button>
            }
          </div>
        }
        <div class="bottom-space"></div>
      </div>
    </ion-content>
  `
})
export class ClientSettingsPage implements OnInit {
  private readonly clientApi = inject(ClientApiService);
  private readonly alertController = inject(AlertController);
  private readonly toastController = inject(ToastController);

  me: ClientMeResponse | null = null;
  pendingRequest: ClientDetailChangeRequest | null = null;
  deletionRequest: ClientDetailChangeRequest | null = null;
  isEditing = false;
  isSubmittingEdit = false;
  draftAnswers: Record<string, string> = {};
  editNote = '';
  currentPassword = '';
  newPassword = '';
  confirmPassword = '';
  isChangingPassword = false;
  message = '';
  messageIsError = false;

  get initials(): string {
    const client = this.me?.client;
    return `${client?.first_name?.[0] || ''}${client?.last_name?.[0] || ''}`.toUpperCase() || 'C';
  }

  get editableFields(): Array<{ key: string; label: string }> {
    const fields = (this.me?.registration_fields || []) as DynamicField[];
    const core = [
      { key: 'first_name', label: 'First name' },
      { key: 'last_name', label: 'Last name' },
      { key: 'email', label: 'Email' }
    ];
    const custom = fields
      .filter((field) => !field.is_core)
      .map((field) => ({ key: field.key || field.label, label: field.label }));
    return [...core, ...custom];
  }

  get answeredFields(): Array<{ key: string; label: string; value: string }> {
    const answers = this.me?.client?.registration_answers || {};
    const skip = new Set(['first_name', 'last_name', 'email']);
    const fields = (this.me?.registration_fields || []) as DynamicField[];

    return Object.entries(answers)
      .filter(([key, value]) => !skip.has(key) && String(value ?? '').trim() !== '')
      .map(([key, value]) => {
        const field = fields.find((item) => (item.key || item.label) === key);
        return { key, label: field?.label || key.replace(/_/g, ' '), value: String(value) };
      });
  }

  ngOnInit(): void {
    this.load();
  }

  onPhoto(event: Event): void {
    const input = event.target as HTMLInputElement;
    const file = input.files?.[0];

    if (!file) {
      return;
    }

    const reader = new FileReader();
    reader.onload = () => {
      this.clientApi.updatePhoto(String(reader.result)).subscribe({
        next: (response) => {
          if (this.me) {
            this.me.client = response.client;
          }

          this.setMessage('Photo updated.', false);
        },
        error: () => this.setMessage('Photo could not be updated.', true)
      });
    };
    reader.readAsDataURL(file);
  }

  startEdit(): void {
    const answers = this.me?.client?.registration_answers || {};
    this.draftAnswers = {};

    for (const field of this.editableFields) {
      const clientValue =
        field.key === 'first_name'
          ? this.me?.client?.first_name
          : field.key === 'last_name'
            ? this.me?.client?.last_name
            : field.key === 'email'
              ? this.me?.client?.email
              : answers[field.key];
      this.draftAnswers[field.key] = String(clientValue ?? '');
    }

    this.editNote = '';
    this.isEditing = true;
  }

  submitEdit(): void {
    this.isSubmittingEdit = true;
    const proposed: Record<string, string> = {};
    Object.entries(this.draftAnswers).forEach(([key, value]) => {
      proposed[key] = String(value ?? '');
    });

    this.clientApi.submitDetailChangeRequest(proposed, this.editNote.trim()).subscribe({
      next: (response) => {
        this.pendingRequest = response.change_request;
        this.isSubmittingEdit = false;
        this.isEditing = false;
        this.setMessage(response.message || 'Edit request sent to your trainer.', false);
      },
      error: () => {
        this.isSubmittingEdit = false;
        this.setMessage('Edit request could not be sent.', true);
      }
    });
  }

  changePassword(): void {
    this.isChangingPassword = true;
    this.clientApi.changePassword(this.currentPassword, this.newPassword, this.confirmPassword).subscribe({
      next: (response) => {
        // The backend rotates the token on password change - keep the
        // session alive by storing the fresh one.
        if (response.token) {
          this.clientApi.storeSession(response.token, response.client);
        }

        this.isChangingPassword = false;
        this.currentPassword = '';
        this.newPassword = '';
        this.confirmPassword = '';
        this.setMessage(response.message || 'Password updated.', false);
      },
      error: () => {
        this.isChangingPassword = false;
        this.setMessage('Password change failed. Check your current password and strength.', true);
      }
    });
  }

  async requestDeletion(): Promise<void> {
    const alert = await this.alertController.create({
      header: 'Request account deletion?',
      message: 'Your trainer will review this request. Your data stays until it is approved.',
      inputs: [{ name: 'note', type: 'textarea', placeholder: 'Optional note to your trainer' }],
      buttons: [
        { text: 'Cancel', role: 'cancel' },
        {
          text: 'Request deletion',
          role: 'destructive',
          handler: (values: { note?: string }) => {
            this.clientApi.requestAccountDeletion((values.note || '').trim()).subscribe({
              next: (response) => {
                this.deletionRequest = response.deletion_request;
                this.setMessage(response.message || 'Deletion request sent.', false);
              },
              error: () => this.setMessage('Deletion request could not be sent.', true)
            });
          }
        }
      ]
    });
    await alert.present();
  }

  withdrawDeletion(): void {
    this.clientApi.withdrawAccountDeletionRequest().subscribe({
      next: (response) => {
        this.deletionRequest = null;
        this.setMessage(response.message || 'Deletion request withdrawn.', false);
      },
      error: () => this.setMessage('Could not withdraw the request.', true)
    });
  }

  private load(): void {
    this.clientApi.getMe().subscribe({
      next: (me) => (this.me = me),
      error: () => this.setMessage('Could not load your profile.', true)
    });
    this.clientApi.getDetailChangeRequest().subscribe({
      next: (response) => (this.pendingRequest = response.change_request?.status === 'pending' ? response.change_request : null),
      error: () => undefined
    });
    this.clientApi.getAccountDeletionRequest().subscribe({
      next: (response) => (this.deletionRequest = response.deletion_request),
      error: () => undefined
    });
  }

  private setMessage(text: string, isError: boolean): void {
    this.message = text;
    this.messageIsError = isError;
    setTimeout(() => (this.message = ''), 4000);
  }

  private async toast(text: string): Promise<void> {
    const toast = await this.toastController.create({ message: text, duration: 1800, position: 'bottom' });
    await toast.present();
  }
}

File: mobile_flutter/lib/features/client/client_settings_page.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import 'package:image_picker/image_picker.dart';

import '../../app/router.dart';
import '../../core/api/api_client.dart';
import '../../core/api/client_api.dart';
import '../../core/api/models/forms_groups_models.dart';
import '../../core/config/env.dart';
import '../../core/theme/app_tokens.dart';
import '../../shared/widgets/app_widgets.dart';
import '../../shared/widgets/password_field.dart';
import '../trainer/trainer_format.dart';

/// Client settings — profile, trainer-approved edit requests, password, and
/// account deletion requests.
/// Replica of mobile/src/app/pages/client/settings/client-settings.page.ts.
class ClientSettingsPage extends ConsumerStatefulWidget {
  const ClientSettingsPage({super.key});

  @override
  ConsumerState<ClientSettingsPage> createState() => _ClientSettingsPageState();
}

class _ClientSettingsPageState extends ConsumerState<ClientSettingsPage> {
  ClientMeResponse? _me;
  ClientDetailChangeRequest? _pendingRequest;
  ClientDetailChangeRequest? _deletionRequest;
  bool _loading = true;

  bool _isEditing = false;
  bool _isSubmittingEdit = false;
  final Map<String, TextEditingController> _draft = {};
  final _editNote = TextEditingController();

  final _currentPassword = TextEditingController();
  final _newPassword = TextEditingController();
  final _confirmPassword = TextEditingController();
  bool _isChangingPassword = false;

  String _message = '';
  bool _messageIsError = false;

  @override
  void initState() {
    super.initState();
    _load();
  }

  @override
  void dispose() {
    for (final c in _draft.values) {
      c.dispose();
    }
    _editNote.dispose();
    _currentPassword.dispose();
    _newPassword.dispose();
    _confirmPassword.dispose();
    super.dispose();
  }

  String get _initials {
    final client = _me?.client;
    final first = client?.firstName ?? '';
    final last = client?.lastName ?? '';
    final letters =
        '${first.isNotEmpty ? first[0] : ''}${last.isNotEmpty ? last[0] : ''}'
            .toUpperCase();
    return letters.isEmpty ? 'C' : letters;
  }

  /// Core identity fields plus the group's custom registration fields.
  List<({String key, String label})> get _editableFields {
    final custom = (_me?.registrationFields ?? [])
        .where((field) => !field.isCore)
        .map((field) => (key: field.answerKey, label: field.label));
    return [
      (key: 'first_name', label: 'First name'),
      (key: 'last_name', label: 'Last name'),
      (key: 'email', label: 'Email'),
      ...custom,
    ];
  }

  /// Answered custom fields for the read-only view — identity is shown above.
  List<({String label, String value})> get _answeredFields {
    final answers = _me?.client.registrationAnswers ?? const {};
    const skip = {'first_name', 'last_name', 'email'};
    final fields = _me?.registrationFields ?? const <DynamicField>[];

    return answers.entries
        .where((e) => !skip.contains(e.key) && e.value.trim().isNotEmpty)
        .map((e) {
      final field = fields.where((f) => f.answerKey == e.key).firstOrNull;
      return (
        label: field?.label ?? e.key.replaceAll('_', ' '),
        value: e.value,
      );
    }).toList();
  }

  Future<void> _load() async {
    final api = ref.read(clientApiProvider);
    try {
      final me = await api.getMe();
      if (mounted) setState(() => _me = me);
    } catch (_) {
      _setMessage('Could not load your profile.', true);
    }
    try {
      final request = await api.getDetailChangeRequest();
      if (mounted) {
        // Only a pending request is actionable; approved/rejected are history.
        setState(() =>
            _pendingRequest = request?.status == 'pending' ? request : null);
      }
    } catch (_) {}
    try {
      final deletion = await api.getAccountDeletionRequest();
      if (mounted) setState(() => _deletionRequest = deletion);
    } catch (_) {}
    if (mounted) setState(() => _loading = false);
  }

  void _setMessage(String text, bool isError) {
    if (!mounted) return;
    setState(() {
      _message = text;
      _messageIsError = isError;
    });
    Future.delayed(const Duration(seconds: 4), () {
      if (mounted) setState(() => _message = '');
    });
  }

  Future<void> _pickPhoto() async {
    final picked = await ImagePicker().pickImage(
      source: ImageSource.gallery,
      maxWidth: 1000,
      imageQuality: 85,
    );
    if (picked == null) return;

    try {
      // The endpoint takes a data URL, same as the TS FileReader result.
      final bytes = await File(picked.path).readAsBytes();
      final ext = picked.name.split('.').last.toLowerCase();
      final mime = ext == 'png' ? 'image/png' : 'image/jpeg';
      final dataUrl = 'data:$mime;base64,${base64Encode(bytes)}';

      final client = await ref.read(clientApiProvider).updatePhoto(dataUrl);
      if (!mounted) return;
      setState(() {
        final me = _me;
        if (me != null) {
          _me = ClientMeResponse(
            client: client,
            group: me.group,
            registrationFields: me.registrationFields,
            sharedAdditionalInfo: me.sharedAdditionalInfo,
            trainerProfile: me.trainerProfile,
          );
        }
      });
      _setMessage('Photo updated.', false);
    } on ApiException catch (error) {
      _setMessage(error.message, true);
    } catch (_) {
      _setMessage('Photo could not be updated.', true);
    }
  }

  void _startEdit() {
    final client = _me?.client;
    final answers = client?.registrationAnswers ?? const {};

    for (final c in _draft.values) {
      c.dispose();
    }
    _draft.clear();

    for (final field in _editableFields) {
      final value = switch (field.key) {
        'first_name' => client?.firstName ?? '',
        'last_name' => client?.lastName ?? '',
        'email' => client?.email ?? '',
        _ => answers[field.key] ?? '',
      };
      _draft[field.key] = TextEditingController(text: value);
    }
    _editNote.clear();
    setState(() => _isEditing = true);
  }

  Future<void> _submitEdit() async {
    setState(() => _isSubmittingEdit = true);
    final proposed = {
      for (final entry in _draft.entries) entry.key: entry.value.text,
    };

    try {
      final request = await ref
          .read(clientApiProvider)
          .submitDetailChangeRequest(proposed, note: _editNote.text.trim());
      if (!mounted) return;
      setState(() {
        _pendingRequest = request;
        _isSubmittingEdit = false;
        _isEditing = false;
      });
      _setMessage('Edit request sent to your trainer.', false);
    } on ApiException catch (error) {
      if (!mounted) return;
      setState(() => _isSubmittingEdit = false);
      _setMessage(error.message, true);
    }
  }

  bool get _canChangePassword =>
      !_isChangingPassword &&
      _currentPassword.text.isNotEmpty &&
      _newPassword.text.length >= 8 &&
      _newPassword.text == _confirmPassword.text;

  Future<void> _changePassword() async {
    setState(() => _isChangingPassword = true);
    final api = ref.read(clientApiProvider);
    try {
      final response = await api.changePassword(
        _currentPassword.text,
        _newPassword.text,
        _confirmPassword.text,
      );
      // The backend rotates the token — store the fresh one or the session dies.
      final client = response.client;
      if (response.token.isNotEmpty && client != null) {
        await api.storeSession(response.token, client);
      }
      if (!mounted) return;
      setState(() {
        _isChangingPassword = false;
        _currentPassword.clear();
        _newPassword.clear();
        _confirmPassword.clear();
      });
      _setMessage('Password updated.', false);
    } on ApiException catch (error) {
      if (!mounted) return;
      setState(() => _isChangingPassword = false);
      _setMessage(error.message, true);
    }
  }

  Future<void> _requestDeletion() async {
    final note = TextEditingController();
    final confirmed = await showDialog<bool>(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Request account deletion?'),
        content: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text(
              'Your trainer will review this request. Your data stays until it '
              'is approved.',
              style: context.text.bodySmall,
            ),
            const SizedBox(height: AppSpacing.md),
            TextField(
              controller: note,
              maxLines: 3,
              decoration: const InputDecoration(
                labelText: 'Note to your trainer (optional)',
              ),
            ),
          ],
        ),
        actions: [
          TextButton(onPressed: () => context.pop(false), child: const Text('Cancel')),
          FilledButton(
            onPressed: () => context.pop(true),
            style: FilledButton.styleFrom(
              backgroundColor: context.colors.error,
              minimumSize: const Size(0, AppSize.buttonHeightSm),
            ),
            child: const Text('Request deletion'),
          ),
        ],
      ),
    );

    final text = note.text.trim();
    note.dispose();
    if (confirmed != true) return;

    try {
      final request = await ref
          .read(clientApiProvider)
          .requestAccountDeletion(note: text);
      if (mounted) setState(() => _deletionRequest = request);
      _setMessage('Deletion request sent.', false);
    } on ApiException catch (error) {
      _setMessage(error.message, true);
    }
  }

  Future<void> _withdrawDeletion() async {
    try {
      await ref.read(clientApiProvider).withdrawAccountDeletionRequest();
      if (mounted) setState(() => _deletionRequest = null);
      _setMessage('Deletion request withdrawn.', false);
    } on ApiException catch (error) {
      _setMessage(error.message, true);
    }
  }

  @override
  Widget build(BuildContext context) {
    if (_loading) {
      return Scaffold(
        appBar: AppBar(title: const Text('Settings')),
        body: const PagePad(children: [SkeletonBox(height: 110), SkeletonBox(height: 200)]),
      );
    }

    final client = _me?.client;
    final tokens = context.tokens;

    return Scaffold(
      appBar: AppBar(
        title: const Text('Settings'),
        leading: BackButton(onPressed: () => context.go(Routes.clientMore)),
      ),
      body: PagePad(
        onRefresh: _isEditing ? null : _load,
        children: [
          if (_message.isNotEmpty)
            Padding(
              padding: const EdgeInsets.only(bottom: AppSpacing.md),
              child: AppCard(
                color: (_messageIsError ? context.colors.error : tokens.success)
                    .withValues(alpha: 0.08),
                child: Text(
                  _message,
                  style: context.text.bodySmall?.copyWith(
                    color: _messageIsError ? context.colors.error : tokens.success,
                  ),
                ),
              ),
            ),

          AppCard(
            child: Row(
              children: [
                AppAvatar(
                  initials: _initials,
                  imageUrl: Env.mediaUrl(client?.photo ?? ''),
                  size: 56,
                ),
                const SizedBox(width: AppSpacing.md),
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        client?.displayName ?? '',
                        style: context.text.titleMedium,
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                      ),
                      Text(client?.username ?? '', style: context.text.bodySmall),
                      const SizedBox(height: AppSpacing.xs),
                      OutlinedButton.icon(
                        onPressed: _pickPhoto,
                        icon: const Icon(Icons.photo_camera_outlined,
                            size: AppSize.iconRow),
                        label: const Text('Change photo'),
                        style: OutlinedButton.styleFrom(
                          minimumSize: const Size(0, AppSize.buttonHeightSm),
                        ),
                      ),
                    ],
                  ),
                ),
              ],
            ),
          ),

          if (_pendingRequest != null) ...[
            const SizedBox(height: AppSpacing.md),
            AppCard(
              color: tokens.primarySoft,
              child: Row(
                children: [
                  Icon(Icons.hourglass_top_outlined,
                      size: 20, color: context.colors.primary),
                  const SizedBox(width: AppSpacing.sm),
                  Expanded(
                    child: Text(
                      'Your edit request is waiting for your trainer to review.',
                      style: context.text.bodySmall,
                    ),
                  ),
                ],
              ),
            ),
          ],

          if (!_isEditing) ...[
            SectionHeader(
              title: 'Your details',
              actionLabel: _pendingRequest == null ? 'Request edit' : null,
              onAction: _pendingRequest == null ? _startEdit : null,
            ),
            AppCard(
              child: Column(
                children: [
                  _kv('Name', client?.displayName ?? ''),
                  _kv('Email', client?.email ?? ''),
                  _kv('Username', client?.username ?? ''),
                  _kv('Group', client?.groupName ?? ''),
                  for (final field in _answeredFields) _kv(field.label, field.value),
                ],
              ),
            ),
            Padding(
              padding: const EdgeInsets.only(top: AppSpacing.sm),
              child: Text(
                'Changes to your details need your trainer to approve them.',
                style: context.text.bodySmall,
              ),
            ),
          ] else ...[
            const SectionHeader(title: 'Request an edit'),
            AppCard(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    'Your trainer reviews these changes before they apply.',
                    style: context.text.bodySmall,
                  ),
                  const SizedBox(height: AppSpacing.md),
                  for (final field in _editableFields)
                    Padding(
                      padding: const EdgeInsets.only(bottom: AppSpacing.md),
                      child: TextField(
                        controller: _draft[field.key],
                        decoration: InputDecoration(labelText: field.label),
                      ),
                    ),
                  TextField(
                    controller: _editNote,
                    maxLines: 2,
                    decoration: const InputDecoration(
                      labelText: 'Note to your trainer (optional)',
                    ),
                  ),
                  const SizedBox(height: AppSpacing.md),
                  Row(
                    children: [
                      Expanded(
                        child: FilledButton(
                          onPressed: _isSubmittingEdit ? null : _submitEdit,
                          child: Text(
                            _isSubmittingEdit ? 'Sending…' : 'Send request',
                          ),
                        ),
                      ),
                      const SizedBox(width: AppSpacing.sm),
                      TextButton(
                        onPressed: () => setState(() => _isEditing = false),
                        child: const Text('Cancel'),
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ],

          const SectionHeader(title: 'Password'),
          AppCard(
            child: Column(
              children: [
                PasswordField(
                  controller: _currentPassword,
                  label: 'Current password',
                  textInputAction: TextInputAction.next,
                ),
                const SizedBox(height: AppSpacing.md),
                PasswordField(
                  controller: _newPassword,
                  label: 'New password',
                  textInputAction: TextInputAction.next,
                  autofillHints: const [AutofillHints.newPassword],
                ),
                const SizedBox(height: AppSpacing.md),
                PasswordField(
                  controller: _confirmPassword,
                  label: 'Confirm new password',
                  autofillHints: const [AutofillHints.newPassword],
                ),
                const SizedBox(height: AppSpacing.md),
                FilledButton(
                  onPressed: _canChangePassword ? _changePassword : null,
                  child: Text(
                    _isChangingPassword ? 'Updating…' : 'Change password',
                  ),
                ),
              ],
            ),
          ),

          const SectionHeader(title: 'Account'),
          AppCard(
            child: _deletionRequest != null
                ? Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Row(
                        children: [
                          Icon(Icons.hourglass_top_outlined,
                              size: 20, color: context.colors.error),
                          const SizedBox(width: AppSpacing.sm),
                          Expanded(
                            child: Text(
                              'Deletion requested ${shortDate(_deletionRequest!.createdAt)}',
                              style: context.text.titleSmall,
                            ),
                          ),
                        ],
                      ),
                      const SizedBox(height: AppSpacing.xs),
                      Text(
                        'Your trainer is reviewing it. Your data stays until '
                        'they approve.',
                        style: context.text.bodySmall,
                      ),
                      const SizedBox(height: AppSpacing.md),
                      OutlinedButton(
                        onPressed: _withdrawDeletion,
                        style: OutlinedButton.styleFrom(
                          minimumSize: const Size(0, AppSize.buttonHeightSm),
                        ),
                        child: const Text('Withdraw request'),
                      ),
                    ],
                  )
                : Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text('Delete your account', style: context.text.titleSmall),
                      const SizedBox(height: AppSpacing.xs),
                      Text(
                        'Your trainer reviews the request. Nothing is removed '
                        'until they approve it.',
                        style: context.text.bodySmall,
                      ),
                      const SizedBox(height: AppSpacing.md),
                      OutlinedButton(
                        onPressed: _requestDeletion,
                        style: OutlinedButton.styleFrom(
                          foregroundColor: context.colors.error,
                          side: BorderSide(color: context.colors.error),
                          minimumSize: const Size(0, AppSize.buttonHeightSm),
                        ),
                        child: const Text('Request deletion'),
                      ),
                    ],
                  ),
          ),
        ],
      ),
    );
  }

  Widget _kv(String label, String value) {
    return Padding(
      padding: const EdgeInsets.only(bottom: AppSpacing.sm),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Expanded(flex: 2, child: Text(label, style: context.text.bodySmall)),
          Expanded(
            flex: 3,
            child: Text(
              value.trim().isEmpty ? '—' : value,
              style: context.text.titleSmall,
              textAlign: TextAlign.right,
            ),
          ),
        ],
      ),
    );
  }
}
